jwt: builder panics when using custom struct as claims

Builder panics when you use a custom claim struct.

```go
package main

import (
	"crypto/rand"
	"crypto/rsa"
	"log"

	"gopkg.in/square/go-jose.v2"
	"gopkg.in/square/go-jose.v2/jwt"
)

type claims struct {
	Subject string `json:"sub"`
}

func main() {
	priv, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		log.Fatalf("generate: %v", err)
	}
	signingKey := jose.SigningKey{Algorithm: jose.RS256, Key: priv}
	signer, err := jose.NewSigner(signingKey, nil)
	if err != nil {
		log.Fatalf("generate: %v", err)
	}
	token, err := jwt.Signed(signer).Claims(claims{"foo"}).CompactSerialize()
	if err != nil {
		log.Fatalf("creating jwt: %v", err)
	}
	log.Println(token)
}
```

Running this program results in.

```
$ go run main.go 
panic: reflect: call of reflect.Value.IsNil on struct Value

goroutine 1 [running]:
panic(0x51c940, 0xc420196e20)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
reflect.Value.IsNil(0x5216a0, 0xc42007c4a0, 0x99, 0xc42007c4a0)
	/usr/local/go/src/reflect/value.go:977 +0xf9
gopkg.in/square/go-jose.v2/jwt.publicClaims(0x5216a0, 0xc42007c4a0, 0x0, 0x18, 0x0)
	/home/eric/src/gopkg.in/square/go-jose.v2/jwt/json.go:103 +0x70
gopkg.in/square/go-jose.v2/jwt.marshalClaims(0x5216a0, 0xc42007c4a0, 0x8, 0x10, 0xc42007c4a0, 0x0, 0x5dec40)
	/home/eric/src/gopkg.in/square/go-jose.v2/jwt/json.go:125 +0x96
gopkg.in/square/go-jose.v2/jwt.(*Builder).Claims(0xc420078320, 0x5216a0, 0xc42007c4a0, 0x5216a0)
	/home/eric/src/gopkg.in/square/go-jose.v2/jwt/builder.go:79 +0x6e
main.main()
	/home/eric/p/go/jose/main.go:26 +0x176
exit status 2
```

--- a/jwt/builder_test.go
+++ b/jwt/builder_test.go
@@ -86,7 +86,7 @@ func TestBuilderHeadersSigner(t *testing.T) {
 		}
 		jws, err := jose.ParseSigned(token)
 		if err != nil {
-			t.Errorf("case %d: parse signed: %v", err)
+			t.Errorf("case %d: parse signed: %v", i, err)
 			continue
 		}
 		gotKeyIDs := make([]string, len(jws.Signatures))
@@ -96,7 +96,7 @@ func TestBuilderHeadersSigner(t *testing.T) {
 		sort.Strings(wantKeyIDs)
 		sort.Strings(gotKeyIDs)
 		if !reflect.DeepEqual(wantKeyIDs, gotKeyIDs) {
-			t.Errorf("case %d: wanted=%q got=%q", wantKeyIDs, gotKeyIDs)
+			t.Errorf("case %d: wanted=%q got=%q", i, wantKeyIDs, gotKeyIDs)
 		}
 	}
 }

--- a/jwt/claims_test.go
+++ b/jwt/claims_test.go
@@ -18,11 +18,68 @@
 package jwt
 
 import (
+	"crypto/rand"
+	"crypto/rsa"
 	"testing"
 	"time"
 
 	"github.com/stretchr/testify/assert"
+	"gopkg.in/square/go-jose.v2"
 )
+
+type testClaims struct {
+	Subject string `json:"sub"`
+}
+
+func TestCustomClaimsNonPointer(t *testing.T) {
+	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		t.Fatalf("error generating key: %v", err)
+	}
+	signingKey := jose.SigningKey{Algorithm: jose.RS256, Key: priv}
+	signer, err := jose.NewSigner(signingKey, nil)
+	if err != nil {
+		t.Fatalf("error generating token: %v", err)
+	}
+	jwt, err := Signed(signer).Claims(testClaims{"foo"}).CompactSerialize()
+	if err != nil {
+		t.Fatalf("error creating jwt: %v", err)
+	}
+	parsed, err := ParseSigned(jwt)
+	if err != nil {
+		t.Fatalf("error parsing jwt: %v", err)
+	}
+
+	out := &testClaims{}
+	if assert.NoError(t, parsed.Claims(out, &priv.PublicKey)) {
+		assert.Equal(t, out.Subject, "foo")
+	}
+}
+
+func TestCustomClaimsPointer(t *testing.T) {
+	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		t.Fatalf("error generating key: %v", err)
+	}
+	signingKey := jose.SigningKey{Algorithm: jose.RS256, Key: priv}
+	signer, err := jose.NewSigner(signingKey, nil)
+	if err != nil {
+		t.Fatalf("error generating token: %v", err)
+	}
+	jwt, err := Signed(signer).Claims(&testClaims{"foo"}).CompactSerialize()
+	if err != nil {
+		t.Fatalf("error creating jwt: %v", err)
+	}
+	parsed, err := ParseSigned(jwt)
+	if err != nil {
+		t.Fatalf("error parsing jwt: %v", err)
+	}
+
+	out := &testClaims{}
+	if assert.NoError(t, parsed.Claims(out, &priv.PublicKey)) {
+		assert.Equal(t, out.Subject, "foo")
+	}
+}
 
 func TestEncodeClaims(t *testing.T) {
 	now := time.Date(2016, 1, 1, 0, 0, 0, 0, time.UTC)

--- a/jwt/claims_test.go
+++ b/jwt/claims_test.go
@@ -18,8 +18,6 @@
 package jwt
 
 import (
-	"crypto/rand"
-	"crypto/rsa"
 	"testing"
 	"time"
 
@@ -32,11 +30,7 @@ type testClaims struct {
 }
 
 func TestCustomClaimsNonPointer(t *testing.T) {
-	priv, err := rsa.GenerateKey(rand.Reader, 2048)
-	if err != nil {
-		t.Fatalf("error generating key: %v", err)
-	}
-	signingKey := jose.SigningKey{Algorithm: jose.RS256, Key: priv}
+	signingKey := jose.SigningKey{Algorithm: jose.RS256, Key: testPrivRSAKey1}
 	signer, err := jose.NewSigner(signingKey, nil)
 	if err != nil {
 		t.Fatalf("error generating token: %v", err)
@@ -51,17 +45,13 @@ func TestCustomClaimsNonPointer(t *testing.T) {
 	}
 
 	out := &testClaims{}
-	if assert.NoError(t, parsed.Claims(out, &priv.PublicKey)) {
+	if assert.NoError(t, parsed.Claims(out, &testPrivRSAKey1.PublicKey)) {
 		assert.Equal(t, out.Subject, "foo")
 	}
 }
 
 func TestCustomClaimsPointer(t *testing.T) {
-	priv, err := rsa.GenerateKey(rand.Reader, 2048)
-	if err != nil {
-		t.Fatalf("error generating key: %v", err)
-	}
-	signingKey := jose.SigningKey{Algorithm: jose.RS256, Key: priv}
+	signingKey := jose.SigningKey{Algorithm: jose.RS256, Key: testPrivRSAKey1}
 	signer, err := jose.NewSigner(signingKey, nil)
 	if err != nil {
 		t.Fatalf("error generating token: %v", err)
@@ -76,7 +66,7 @@ func TestCustomClaimsPointer(t *testing.T) {
 	}
 
 	out := &testClaims{}
-	if assert.NoError(t, parsed.Claims(out, &priv.PublicKey)) {
+	if assert.NoError(t, parsed.Claims(out, &testPrivRSAKey1.PublicKey)) {
 		assert.Equal(t, out.Subject, "foo")
 	}
 }

--- a/jwt/json.go
+++ b/jwt/json.go
@@ -99,12 +99,11 @@ func (s *audience) UnmarshalJSON(b []byte) error {
 var claimsType = reflect.TypeOf((*Claims)(nil)).Elem()
 
 func publicClaims(cl interface{}) (*Claims, error) {
-	v := reflect.ValueOf(cl)
-	if v.IsNil() || v.Kind() != reflect.Ptr || v.Elem().Kind() != reflect.Struct {
+	v := reflect.Indirect(reflect.ValueOf(cl))
+	if v.Kind() != reflect.Struct {
 		return nil, ErrInvalidClaims
 	}
 
-	v = v.Elem()
 	f := v.FieldByName("Claims")
 	if !f.IsValid() || f.Type() != claimsType {
 		return nil, nil

